README
	       Off-the-Record Messaging plugin for libpurple
			    v3.2.1, 19 Dec 2011
		  based on pidgin-otr v3.2.0, 15 Jun 2008

This is a purple plugin which implements Off-the-Record (OTR) Messaging.
It is known to work (at least) under the Linux and Windows versions of
pidgin (2.10.1+). Note that patches are required for pidgin/finch 2.10.1
to get proper functionality.

OTR allows you to have private conversations over IM by providing:
 - Encryption
   - No one else can read your instant messages.
 - Authentication
   - You are assured the correspondent is who you think it is.
 - Deniability
   - The messages you send do _not_ have digital signatures that are
     checkable by a third party.  Anyone can forge messages after a
     conversation to make them look like they came from you.  However,
     _during_ a conversation, your correspondent is assured the messages
     he sees are authentic and unmodified.
 - Perfect forward secrecy
   - If you lose control of your private keys, no previous conversation
     is compromised.

For more information on Off-the-Record Messaging, see
http://otr.cypherpunks.ca/

USAGE

Run pidgin or finch, and open the Plugins panel.  Find the
Off-the-Record Messaging plugin, and enable it by selecting the
checkbox next to it.  That should be all you need to do. If the
plugin is not in the list, you may need to use the "Install Plugin..."
button to locate the plugin file. It should normally be in
/usr/lib/purple-2/purple-otr.so.


CONFIGURATION

Click "Configure Plugin" to bring up the plugin settings.

The OTR settings determine when private messaging is enabled.  The
checkboxes on this page control the default settings; you can edit
the per-buddy settings by right-clicking on your buddy in the buddy
list, and choosing "OTR Settings" from the menu.

The settings are:
    [X] Enable private messaging
    [X] Automatically initiate private messaging
    [ ] Require private messaging
    [ ] Don't log OTR conversations

    If the "enable private messaging" box is unchecked, private messages
    will be disabled completely (and the other boxes will be ignored,
    as they're irrelevant).

    If the first box is checked, but "automatically initiate private
    messaging" is unchecked, private messaging will be enabled, but only
    if either you or your buddy explicitly requests to start a private
    conversation (and the third box will be ignored, as it's
    irrelevant).

    If the first two boxes are checked, but "require private messaging"
    is unchecked, OTR will attempt to detect whether your buddy can
    understand OTR private messages, and if so, automatically start a
    private conversation.

    If the first three boxes are checked, messages will not be sent to your
    buddy unless you are in a private conversation.

    If the fourth box is checked, OTR-protected conversations will not
    be logged, even if logging of instant messages is turned on in
    pidgin.


KEY MANAGEMENT

The main "Plugins" menu in Finch and the "Tools" menu in Pidgin
(available when viewing the Buddy List) will also have a submenu for
"Off-the-Record Messaging". This submenu has two options, "Get Private
Key" and "Manage Fingerprints".

Get Private Keys:
    Private keys are used to authenticate you to your buddies.  OTR will
    automatically generate private keys when needed, but you can also
    generate them manually if you wish by using the "Get Key" button
    here.  Choose one of your accounts from the menu, click "Get Key"
    and wait until it's finished.  You'll see a new "Private Key" dialog
    with a sequence of letters when it's done. (If there is already a
    private key for the account; it will simply be displayed, no new key
    will be generated.) This is the "fingerprint" for that account; it is
    unique to that account.  If you have multiple IM accounts, you can
    generate private keys for each one separately.

Manage Fingerprints:
    This dialog allows you to see the fingerprints of any
    buddies you have previously communicated with privately.

    The "Status" will indicate the current OTR status of any
    conversation using each fingerprint.  The possibilities are
    "Private", which means you're having a private conversation,
    "Unverified", which means you have not yet verified your buddy's
    fingerprint, "Not private", which means you're just chatting in IM
    the usual (non-OTR) way, and "Finished", which means your buddy has
    selected "Stop Privacy"; at this point, you will be
    unable to send messages to him at all, until you either also choose
    "Stop Privacy" (in which case further messages will be
    sent unencrypted), or else choose "Refresh privacy" (in
    which case further messages will be sent privately).

    The table also indicates whether or not you have verified this
    fingerprint by authenticating your buddy.

    By selecting one of your buddies from the list, you'll be able to do
    one or more of the following things by clicking the buttons below
    the list:
     - "Verify": this will open a window where you can
       verify the value of your buddies' fingerprint.  If you do not
       wish to work with fingerprints directly, you should instead
       authenticate used the OTR menu from within a conversation.
     - "Forget": this will remove your buddy's fingerprint
       from the list.  You'll have to re-authenticate him the next time
       you start a private conversation with him.  Note that you can't
       forget a fingerprint that's currently in use in a private
       conversation.


CONVERSATIONS

IM as normal with your buddies.  If you want to start a private
conversation with one of them, bring up the OTR menu. In Finch it
will be under the "Plugins" menu; in Pidgin it will be under the "Options"
menu.  From the OTR menu, select "Start Privacy". Note that the OTR menu
label will also include text indicating the privacy state of the current
conversation, e.g. "OTR (Not Private)".

If your buddy does not have the OTR plugin, a private conversation will
(of course) not be started.  [But he or she will get some information
about OTR instead.]

If your buddy does have the OTR plugin (and it's enabled), a private
conversation will be initiated.

If both you and your buddy have OTR software, and your OTR settings set
to automatically initiate private messaging, your clients may recognize
each other and automatically start a private conversation.

The first time you have a private conversation with one of your buddies,
a message will appear in your conversation telling you to authenticate
them.  You may authenticate by selecting "Authenticate Buddy" on the
OTR menu.  This is described later on.

At this point, the label on the OTR menu in the conversation window
will change to "OTR (Unverified)".  This means that, although you are
sending encrypted messages, you have not yet authenticated your buddy,
and so it is not certain that the person who can decrypt these messages
is actually your buddy (it may be an attacker).  This situation will
remain until either you or your buddy choose "Authenticate Buddy" from
the OTR menu (described next).

The OTR menu contains the following choices, although some choices
will only be displayed in particular privacy states.:

Start / Refresh Privacy

    Choosing this menu option will attempt to start (or refresh, if
    you're already in one) a private conversation with this buddy.

Stop Privacy

    If you wish to end the private conversation, and go back to
    communicating without privacy protection, you can select this
    option.  Note that if you have "Automatically initiate private
    messaging" set, it is likely that a new private conversation will
    automatically begin immediately.

[Re-] Authenticate Buddy

    For more information on authentication, see
    http://otr-help.cypherpunks.ca/3.2.0/authenticate.php

    OTR provides three ways to authenticate your buddy:

    1) Question and answer
    2) Shared secret
    3) Manual fingerprint verification

    To start the authentication process, you need to first be
    communicating with your buddy in the "Unverified" or "Private"
    states.  [Although the "Private" state indicates that you have
    already successfully authenticated your buddy, and it is not
    necessary to do it again.]  Choose "Authenticate buddy" from the OTR
    menu.  The Authenticate Buddy dialog will pop up.  Use the combo box
    to select which of the three authentication methods you would like
    to use.

    Once you have authenticated your buddy, your OTR status will change
    to "Private".  OTR will also remember that you successfully
    authenticated, and during future private conversations with the same
    buddy, you will no longer get the warning message when you start
    chatting.  This will continue until your buddy switches to a
    computer or an IM account he or she hasn't used before, at which
    point OTR will not recognize him or her and you will be asked to
    authenticate again.

    Question and answer
    -------------------

    To authenticate using a question, pick a question whose answer is
    known only to you and your buddy.  Enter this question and this
    answer, then wait for your buddy to enter the answer too.  If the
    answers don't match, then you may be talking to an imposter.

    If your buddy answers correctly, then you have successfully
    authenticated him or her, and the OTR status of this conversation
    will change to "Private".

    Your buddy will probably also want to ask you a question as well in
    order for him or her to authenticate you back.

    Note that this method first appeared in pidgin-otr 3.2.0; if your
    buddy is using an older version, this will not work.

    Shared secret
    -------------

    To authenticate someone with the shared secret method, you and your
    buddy should decide on a secret word or phrase in advance.  This can
    be done however you like, but you shouldn't type the phrase directly
    into your conversation.

    Enter the shared secret into the field provided in the Authenticate
    Buddy dialog box.  Once you enter the secret and hit OK, your buddy
    will be asked to do exactly the same thing.  If you both enter the
    same text, then OTR will accept that you are really talking to your
    buddy.  Otherwise, OTR reports that authentication has failed.  This
    either means that your buddy made a mistake typing in the text, or
    it may mean that someone is intercepting your communication.

    Note that this method first appeared in pidgin-otr 3.1.0; if your
    buddy is using an older version, this will not work.

    Manual fingerprint verification
    -------------------------------

    If your buddy is using a version of pidgin-otr before 3.1.0, or a
    different OTR client that does not support the other authentication
    methods, you will need to use manual fingerprint verification.

    You will need some other authenticated communication channel (such
    as speaking to your buddy on the telephone, or sending gpg-signed
    messages).  You should tell each other your own fingerprints.  If
    the fingerprint your buddy tells you matches the one listed as his
    or her "purported fingerprint", pull down the selection that says "I
    have not" (verified that this is in fact the correct fingerprint),
    and change it to "I have".

    Once you do this, the OTR status will change to "Private".  Note
    that you only need to do this once per buddy (or once per
    fingerprint, if your buddy has more than one fingerprint).
    pidgin-otr will remember which fingerprints you have marked as
    verified.



NOTES

Please send your bug reports, comments, suggestions, patches, etc. to us
at the contact address below.

This plugin only attempts to protect instant messages, not multi-party
chats, file transfers, etc.

MAILING LISTS

There are three mailing lists pertaining to Off-the-Record Messaging:

otr-announce:
    http://lists.cypherpunks.ca/mailman/listinfo/otr-announce/
    *** All users of OTR software should join this. ***  It is used to
    announce new versions of OTR software, and other important information.

otr-users:
    http://lists.cypherpunks.ca/mailman/listinfo/otr-users/
    Discussion of usage issues related to OTR Messaging software.

otr-dev:
    http://lists.cypherpunks.ca/mailman/listinfo/otr-dev/
    Discussion of OTR Messaging software development.

LICENSE

The Off-the-Record Messaging plugin for pidgin is covered by the following
(GPL) license:

    Off-the-Record Messaging plugin for pidgin
    Copyright (C) 2011 Howard Chu <hyc@symas.com>
    Copyright (C) 2004-2008  Ian Goldberg, Rob Smits,
                             Chris Alexander, Nikita Borisov
                             <otr@cypherpunks.ca>


    This program is free software; you can redistribute it and/or modify
    it under the terms of version 2 of the GNU General Public License as
    published by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    There is a copy of the GNU General Public License in the COPYING file
    packaged with this plugin; if you cannot find it, write to the Free
    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
    02111-1307  USA

CONTACT

To report problems, comments, suggestions, patches, etc., you can email
the authors:

Ian Goldberg, Rob Smits, Chris Alexander, and Nikita Borisov
<otr@cypherpunks.ca>

For more information on Off-the-Record Messaging, visit
http://otr.cypherpunks.ca/


ChangeLog
2011-12-19
	* purple-dialog.[ch]: New code to use libpurple instead of pidgin/GTK
	* purple-ui.[ch]: ditto

	* README: Updated, rewritten for purple-otr 3.2.1.

	* Release 3.2.1.

2008-06-15

	* gtk-dialog.c: Don't try to destroy the submenu when we've
	already destroyed the menu.  Linux doesn't complain about this,
	but Windows does.

	* packaging/windows/pidgin-otr.msi: Bump version number.

	* Makefile.mingw: Add tooltipmenu.o to link line.

	* Release 3.2.0.

2008-06-13

	* po/de.po: Updated German translation from Michael Meier
	<mail@code.mmsources.de>

	* README: Updated, ready for release of 3.2.0.

	* dialogs.h: Remove unused FINGERPRINT_HELPURL.

	* gtk-dialog.c: Remove dead (#if 0) code.

2008-06-11

	* gtk-dialog.c: A small change to aid in i18n from Őry Máté
	<orymate@gmail.com>

	* po/hu.po: Updated Hungarian translation from Őry Máté
	<orymate@gmail.com>

2008-05-30

	* gtk-dialog.c: The new toolbar OTR button now uses the standard
	OTR menu.

	* gtk-dialog.c: The buddy and status items in the OTR menu are
	no longer selectable.

	* gtk-ui.c: Specify that the OTR button shows up in the toolbar
	now.

	* gtk-dialog.c: Added a context-sensitive Help button to the
	Authenticate Buddy dialog.

	* gtk-dialog.c: Revamp of Authenticate Buddy dialog, based on a
	mockup by hcat@gmx.net

2008-05-29

	* gtk-ui.c:
	* gtk-dialog.c: Try moving the OTR button to the toolbar, based
	on an idea from Pontus Andersson at
	http://blog.pesa.se/2007/09/25/goodbye-big-bloated-otr-button/

	* gtk-dialog.c: Some changes to the authenticate buddy dialog;
	more to come.

	* gtk-dialog.c: Some strings were missing _(), and others had
	them unnecessarily.

	* po/de.po: Updated German translation from Michael Meier
	<mail@code.mmsources.de>

2008-05-28

	* tooltipmenu.[ch]: New classes to allow tooltips on menus.

	* otr-icons.h: New OTR icons from <cyrus_xiii@yahoo.com>.

	* README:
	* AUTHORS:
	* *.[ch]: Update copyright dates to 2004-2008 and add Rob Smits
	as an author. (Rob wrote the new gtk-dialog stuff to handle user
	authentication with an explicit question, inserted inline OTR
	icons into the conversation when the status changes, and moved
	the OTR button to the menu bar.)

	* gtk-ui.c: Updated variable names to reflect the (old) change
	from "OTR Options" to "OTR Settings".

	* ui.h:
	* ui.c:
	* gtk-ui.c: Added new OTR UI Options section of the
	configuration panel, which currently controls only whether the
	old OTR button will still show up.

	* dialogs.h:
	* dialogs.c:
	* gtk-dialog.c: 
	* otr-plugin.h:
	* otr-plugin.c: Support starting the SMP authentication with an
	explicit question.

	* otr-plugin.c: Better handling of the case where an SMP message
	fails a verification step.

	* configure.ac:
	* Makefile.mingw: Update version number to 3.2.0.

	* gtk-dialog.c: Made new OTR menus (one master one, and one for
	each active conversation in the window) that replace the old OTR
	button.  [The old button is still available as an option, but
	the menu is always present.]

	* gtk-dialog.c: OTR icons are inserted inline into the
	conversation when the OTR status changes.

	* configure.ac:
	* po/ar.po:
	* po/de.po:
	* po/ru.po: New translations:
	Arabic (Ahmad Gharbeia ﺄﺤﻣﺩ ﻍﺮﺒﻳﺓ <gharbeia@gmail.com>),
	German (Thorsten Mühlfelder <thenktor@gmx.de>),
	Hungarian (Mate Ory <orymate@gmail.com>),
	Russian (Sergei Smirnov <moscow@hro.org>)

	* dialogs.h: Help URLs now have version numbers in them, so the
	appropriate help page can be served.

	* Makefile.am: Include new files in compilation and distribution
	targets.

	* gtk-dialog.c: Added "What's this?" item to OTR menu

	* gtk-dialog.c: Name of authentication dialog is now
	"Authenticate Buddy" with a capital "B".

	* gtk-dialog.c: Authentication dialog now says "Authentication
	Options" instead of "Other Authentication Options".

2007-07-28

	* configure.ac: Fix PKG_CHECK_MODULES so that it gives a
	sensible error if the packages aren't found

	* Makefile.mingw:
	* po/Makefile.mingw: Cleaned up Makefile.mingw files, removed
	"install" target, added "zip" target

	* packaging/windows/pidgin-otr.nsi: Install and uninstall the
	translation files

2007-07-26

	* gtk-dialog.c:
	* otr-plugin.c: ISO C cleanups (no mixing declarations with code)

	* otr-plugin.c: Defend against potential NULL pointer
	dereferences

	* gtk-dialog.c: The SMP secret window should grab focus, but
	only if the user requested to authenticate buddy (and not if it
	popped up because the remote side asked for it)

	* Makefile.mingw: Statically link libintl and libiconv

2007-07-24

	* INSTALL: Added information about i18n; added information about
	linking libgcrypt statically

	* otr-plugin.c: Added support for transparent fragmentation of
	large messages

	* Most files: Updated copyright information

2007-07-22

	* configure.ac: Check for libpurple and use a recent format for
	AC_INIT/AM_INIT_AUTOMAKE, based on a patch from <synx13@sf.net>.

2007-07-18

	* gtk-ui.c: Simplify account list option menu in Config UI,
	thanks to Gabriel Schulhof <nix@go-nix.ca>.

	* gtk-ui.c: Reorder the tabs in the Config UI so that Config is
	first.

	* README: Updated to reflect new tab ordering, and new
	"Don't log OTR conversations" option.

	* gtk-dialog.c: Fixed a bug in the verify fingerprint dialog in
	the unlikely event that a fingerprint exists in the fingerprints
	file with no corresponding key for the local account in the
	private keys file.

2007-07-17

	* dialogs.c:
	* dialogs.h:
	* gtk-dialog.c: Removed Verify fingerprint and View secure
	session id menu options.  Added "Advanced..." button to
	Authenticate buddy dialog, which brings up the old Verify
	fingerprint dialog.  The Authenticate buddy dialog now allows
	users to authenticate their buddies without ever seeing a
	fingerprint.

2007-07-10

	* dialogs.c:
	* dialogs.h:
	* gtk-dialog.c:
	* gtk-ui.c:
	* otr-plugin.c:
	* ui.c:
	* ui.h: Added init() and cleanup() callbacks to the ui and

purple-ui.c
/*
 *  Off-the-Record Messaging plugin for libpurple
 *  Copyright (C) 2011 Howard Chu <hyc@symas.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of version 2 of the GNU General Public License as
 *  published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* config.h */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

/* libgcrypt headers */
#include <gcrypt.h>

/* libotr headers */
#include <libotr/privkey.h>

/* purple headers */
#include "util.h"
#include "account.h"
#include "notify.h"
#include "request.h"

#ifdef ENABLE_NLS
/* internationalisation header */
#include <glib/gi18n-lib.h>
#endif

/* purple-otr headers */
#include "dialogs.h"
#include "ui.h"
#include "otr-plugin.h"

/* Call this function when the DSA key is updated; it will redraw the
 * UI, if visible. */
static void otrg_purple_ui_update_fingerprint(void)
{
    /* Nothing to do */
}

/* Update the keylist, if it's visible */
static void otrg_purple_ui_update_keylist(void)
{
    PurpleNotifySearchResults *res;
    ConnContext *context;
    Fingerprint *fingerprint;
    char hash[45], *item;
    GList *row;
    if (!otrg_keylist_info)
	return;

    res = otrg_keylist_info;
    g_list_foreach(otrg_keylist_strings, (GFunc)g_free, NULL);
    g_list_free(otrg_keylist_strings);
    otrg_keylist_strings = NULL;

    g_list_foreach(res->rows, (GFunc)g_list_free, NULL);
    res->rows = NULL;
    for (context = otrg_plugin_userstate->context_root; context != NULL;
	    context = context->next) {
	int i;
	PurplePlugin *p;
	char *proto_name;
	fingerprint = context->fingerprint_root.next;
	/* If there's no fingerprint, don't add it to the known
	 * fingerprints list */
	while(fingerprint) {
	    row = g_list_append(NULL, context->username);
	    if (context->msgstate == OTRL_MSGSTATE_ENCRYPTED &&
		    context->active_fingerprint != fingerprint) {
		item = _("Unused");
	    } else {
		item = (gchar *)
		    _(otrg_trust_states[otrg_plugin_context_to_trust(context)]);
	    }
	    row = g_list_append(row, item);
	    item = (fingerprint->trust && fingerprint->trust[0]) ?
		_("Yes") : _("No");
	    row = g_list_append(row, item);
	    otrl_privkey_hash_to_human(hash, fingerprint->fingerprint);
	    item = g_strdup(hash);
	    row = g_list_append(row, item);
	    otrg_keylist_strings = g_list_append(otrg_keylist_strings, item);
	    p = purple_find_prpl(context->protocol);
	    proto_name = (p && p->info->name) ? p->info->name : _("Unknown");
	    item = g_strdup_printf("%s (%s)", context->accountname,
		proto_name);
	    row = g_list_append(row, item);
	    otrg_keylist_strings = g_list_append(otrg_keylist_strings, item);
	    purple_notify_searchresults_row_add(res, row);
	    fingerprint = fingerprint->next;
	}
    }
    if (otrg_keylist_handle)
	purple_notify_searchresults_new_rows(NULL, res, otrg_keylist_handle);
}

/* Load the global OTR prefs */
static void otrg_purple_ui_global_prefs_load(gboolean *enabledp,
	gboolean *automaticp, gboolean *onlyprivatep,
	gboolean *avoidloggingotrp)
{
    *enabledp = purple_prefs_get_bool(PREF_ENABLED);
    *automaticp = purple_prefs_get_bool(PREF_AUTO);
    *onlyprivatep = purple_prefs_get_bool(PREF_ONLYPRIV);
    *avoidloggingotrp = purple_prefs_get_bool(PREF_NOLOGOTR);
}

/* Load the OTR prefs for a particular buddy */
static void otrg_purple_ui_buddy_prefs_load(PurpleBuddy *buddy,
	gboolean *usedefaultp, gboolean *enabledp, gboolean *automaticp,
	gboolean *onlyprivatep, gboolean *avoidloggingotrp)
{
    PurpleBlistNode *node = &(buddy->node);

    *usedefaultp = ! purple_blist_node_get_bool(node, "OTR/overridedefault");

    if (*usedefaultp) {
	otrg_purple_ui_global_prefs_load(enabledp, automaticp, onlyprivatep,
		avoidloggingotrp);
    } else {
	*enabledp = purple_blist_node_get_bool(node, _PREF_ENABLED);
	*automaticp = purple_blist_node_get_bool(node, _PREF_AUTO);
	*onlyprivatep = purple_blist_node_get_bool(node, _PREF_ONLYPRIV);
	*avoidloggingotrp =
	    purple_blist_node_get_bool(node, _PREF_NOLOGOTR);
    }
}

/* Save the OTR prefs for a particular buddy */
static void otrg_purple_ui_buddy_prefs_save(PurpleBuddy *buddy,
	gboolean usedefault, gboolean enabled, gboolean automatic,
	gboolean onlyprivate, gboolean avoidloggingotr)
{
    PurpleBlistNode *node = &(buddy->node);

    purple_blist_node_set_bool(node, "OTR/overridedefault", !usedefault);
    purple_blist_node_set_bool(node, _PREF_ENABLED, enabled);
    purple_blist_node_set_bool(node, _PREF_AUTO, automatic);
    purple_blist_node_set_bool(node, _PREF_ONLYPRIV, onlyprivate);
    purple_blist_node_set_bool(node, _PREF_NOLOGOTR, avoidloggingotr);
}

static void config_buddy_cb(void *vbuddy, PurpleRequestFields *fields)
{
    PurpleBuddy *buddy = vbuddy;
    gboolean buddyusedefault, buddyenabled, buddyautomatic,
    	buddyonlyprivate, buddyavoidloggingotr;
    
    buddyusedefault = purple_request_fields_get_bool(fields, "default");
    buddyenabled = purple_request_fields_get_bool(fields, "enabled");
    buddyautomatic = purple_request_fields_get_bool(fields, "auto");
    buddyonlyprivate = purple_request_fields_get_bool(fields, "onlypriv");
    buddyavoidloggingotr = purple_request_fields_get_bool(fields, "nolog");

    /* Apply the changes */
    otrg_purple_ui_buddy_prefs_save(buddy,
    	buddyusedefault, buddyenabled, buddyautomatic,
    	buddyonlyprivate, buddyavoidloggingotr);
}

static void otrg_purple_ui_config_buddy(PurpleBuddy *buddy)
{
    PurpleRequestFields *request;
    PurpleRequestFieldGroup *group;
    PurpleRequestField *field;
    gboolean buddyusedefault, buddyenabled, buddyautomatic,
    	buddyonlyprivate, buddyavoidloggingotr;

    otrg_purple_ui_buddy_prefs_load(buddy, &buddyusedefault, &buddyenabled,
	    &buddyautomatic, &buddyonlyprivate, &buddyavoidloggingotr);

    group = purple_request_field_group_new(NULL);
    field = purple_request_field_bool_new("default", _("Use default "
	    "OTR settings for this buddy"), buddyusedefault);
    purple_request_field_group_add_field(group, field);
    request = purple_request_fields_new();
    purple_request_fields_add_group(request, group);

    group = purple_request_field_group_new(NULL);
    field = purple_request_field_bool_new("enabled", _("Enable private "
	    "messaging"), buddyenabled);
    purple_request_field_group_add_field(group, field);

    field = purple_request_field_bool_new("auto", _("Automatically "
	    "initiate private messaging"), buddyautomatic);
    purple_request_field_group_add_field(group, field);

    field = purple_request_field_bool_new("onlypriv", _("Require private "
	    "messaging"), buddyonlyprivate);
    purple_request_field_group_add_field(group, field);

    field = purple_request_field_bool_new("nolog", _("Don't log OTR conversations"),
	    buddyavoidloggingotr);
    purple_request_field_group_add_field(group, field);

    purple_request_fields_add_group(request, group);

    purple_request_fields(otrg_plugin_handle, _("OTR Settings"), NULL, NULL, request,
    	_("_Set"), G_CALLBACK(config_buddy_cb),
	_("_Cancel"), NULL,
	purple_buddy_get_account(buddy), purple_buddy_get_contact_alias(buddy),
	NULL, buddy);
}

/* Load the preferences for a particular account / username */
static void otrg_purple_ui_get_prefs(OtrgUiPrefs *prefsp, PurpleAccount *account,
	const char *name)
{
    PurpleBuddy *buddy;
    gboolean otrenabled, otrautomatic, otronlyprivate, otravoidloggingotr;
    gboolean buddyusedefault, buddyenabled, buddyautomatic, buddyonlyprivate,
	     buddyavoidloggingotr;

    prefsp->policy = OTRL_POLICY_DEFAULT;
    prefsp->avoid_logging_otr = FALSE;
    prefsp->show_otr_button = FALSE;
    
    /* Get the default policy */
    otrg_purple_ui_global_prefs_load(&otrenabled, &otrautomatic, &otronlyprivate,
	    &otravoidloggingotr);

    if (otrenabled) {
	if (otrautomatic) {
	    if (otronlyprivate) {
		prefsp->policy = OTRL_POLICY_ALWAYS;
	    } else {
		prefsp->policy = OTRL_POLICY_OPPORTUNISTIC;
	    }
	} else {
	    prefsp->policy = OTRL_POLICY_MANUAL;
	}
	prefsp->avoid_logging_otr = otravoidloggingotr;
    } else {
	prefsp->policy = OTRL_POLICY_NEVER;
    }

    buddy = purple_find_buddy(account, name);
    if (!buddy) return;

    /* Get the buddy-specific policy, if present */
    otrg_purple_ui_buddy_prefs_load(buddy, &buddyusedefault, &buddyenabled,
	    &buddyautomatic, &buddyonlyprivate, &buddyavoidloggingotr);

    if (buddyusedefault) return;

    if (buddyenabled) {
	if (buddyautomatic) {
	    if (buddyonlyprivate) {
		prefsp->policy = OTRL_POLICY_ALWAYS;
	    } else {
		prefsp->policy = OTRL_POLICY_OPPORTUNISTIC;
	    }
	} else {
	    prefsp->policy = OTRL_POLICY_MANUAL;
	}
	prefsp->avoid_logging_otr = buddyavoidloggingotr;
    } else {
	prefsp->policy = OTRL_POLICY_NEVER;
    }
}

/* Initialize the OTR UI subsystem */
static void otrg_purple_ui_init(void)
{
    /* Nothing to do */
}

/* Deinitialize the OTR UI subsystem */
static void otrg_purple_ui_cleanup(void)
{
    /* Nothing to do */
}

static const OtrgUiUiOps purple_ui_ui_ops = {
    otrg_purple_ui_init,
    otrg_purple_ui_cleanup,
    otrg_purple_ui_update_fingerprint,
    otrg_purple_ui_update_keylist,
    otrg_purple_ui_config_buddy,
    otrg_purple_ui_get_prefs
};

/* Get the Purple UI ops */
const OtrgUiUiOps *otrg_purple_ui_get_ui_ops(void)
{
    return &purple_ui_ui_ops;
}
